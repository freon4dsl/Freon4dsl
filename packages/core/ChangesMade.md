# Changes made to core and core-svelte

On 19-08-22:
1. Created README in /editor
2. Name of PiActions changed into PiCombinedActions, to distinguish this interface from 
/editor/actions/PiAction. Caused changes in meta, and changes in CustomActions in samples and test.
3. Everything todo with triggers moved from PiActions/PiCombinedActions and PiAction to new file 
/editor/actions/PiTrigger.
4. Moved utils that are specific for the editor from /util to /editor/util, and extracted PiCaret class
from BehaviorUtils.ts. Removed implementation of setCaret in TextBox, because this is always overwritten.
5. Replaced use of 'keyCode' by 'key', only the test whether a character is printable remains to be done.
6. General cleanup of code.
7. Changed 'keyPressAction' to 'isCharAllowed' in TextBox.
8. Cleanup of PiEditor.
9. Merged in the temp changes in tree balancing. 
