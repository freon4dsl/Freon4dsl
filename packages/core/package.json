{
  "name": "@freon4dsl/core",
  "version": "0.7.0-beta2",
  "description": "The uttermost core part of the Freon library, which enables you to build editors for the web",
  "publishConfig": {
    "access": "public"
  },
  "license": "MIT",
  "main": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "type": "module",
  "repository": {
    "type": "git",
    "url": "https://github.com/freon4dsl/Freon4dsl"
  },
  "files": [
    "/dist"
  ],
  "scripts": {
    "build": "rollup -c",
    "build-dev": "rollup -c",
    "clean": "rm -rf node_modules/ && rm -rf dist/",
    "build-release": "npm run build",
    "cleanup": "bash ../../scripts/cleanup.sh",
    "test": "vitest run",
    "start": "echo Start core",
    "release": "npm publish",
    "publish-local": "npm publish --registry http://localhost:4873",
    "unpublish-local": "npm unpublish --force --registry http://localhost:4873",
    "republish-local": "npm run unpublish-local && npm run publish-local"
  },
  "devDependencies": {
    "vitest": "^2.1.0",
    "@vitest/ui": "^2.1.0"
  },
  "dependencies": {
    "lodash": "^4.17.21",
    "@lionweb/validation": "0.6.2",
    "@lionweb/repository-client": "0.2.2",
    "mobx": "^6.12.3",
    "reflect-metadata": "^0.2.2",
    "tslib": "^2.6.3"
  }
}
