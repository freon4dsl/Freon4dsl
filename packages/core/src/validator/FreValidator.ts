import { FreNode } from "../ast";

// Part of the Freon Framework.
export interface FreValidator {
    /**
     * Returns a list of errors on 'modelelement' according to the validation rules
     * stated in the validation definition. If 'includeChildren' is true, the child
     * nodes of 'modelelement' in the AST are also checked.
     *
     * @param modelelement
     * @param includeChildren
     */
    validate(modelelement: FreNode, includeChildren?: boolean): FreError[];
}

/**
 * An error consists of a message coupled to the faulty AST node, either a model
 * element or a list of model elements.
 */
export class FreError {
    message: string;                        // human-readable error message
    reportedOn: FreNode | FreNode[];    // the model element that does not comply
    propertyName: string;                   // the property of the model element that does not comply, if appropriate
    propertyIndex: number;                  // the property index of the model element that does not comply, if appropriate
    locationdescription: string;            // human-readable indication of 'reportedOn'
    severity: FreErrorSeverity;              // indication of how serious the error is, default is 'To Do'
    constructor(message: string, element: FreNode | FreNode[], locationdescription: string, propertyName: string, severity?: FreErrorSeverity, propertyIndex?: number) {
        // todo check whether propertyIndex is really needed
        this.message = message;
        this.reportedOn = element;
        this.locationdescription = locationdescription;
        if (!!severity) {
            this.severity = severity;
        } else {
            this.severity = FreErrorSeverity.ToDo;
        }
        this.propertyName = propertyName;
        this.propertyIndex = propertyIndex;
    }
}

// TODO combine these two Severities
export enum FreErrorSeverity {
    Error = "Error",
    Improvement = "Improvement",
    ToDo = "TODO",
    Info = "Info",
    NONE = "NONE"
}

// severity can range from 0 to 4
// 0 means information
// 1 means hint
// 2 means warning
// 3 means error
export enum SeverityType {
    info = 0,
    hint = 1,
    warning = 2,
    error= 3
}
