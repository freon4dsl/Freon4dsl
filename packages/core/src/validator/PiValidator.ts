import { PiElement } from "../ast";

// Part of the ProjectIt Framework.
export interface PiValidator {
    /**
     * Returns a list of errors on 'modelelement' according to the validation rules
     * stated in the validation definition. If 'includeChildren' is true, the child
     * nodes of 'modelelement' in the AST are also checked.
     *
     * @param modelelement
     * @param includeChildren
     */
    validate(modelelement: PiElement, includeChildren?: boolean): PiError[];
}

/**
 * An error consists of a message coupled to the faulty AST node, either a model
 * element or a list of model elements.
 */
export class PiError {
    message: string;                        // human-readable error message
    reportedOn: PiElement | PiElement[];    // the model element that does not comply
    locationdescription: string;            // human-readable indication of 'reportedOn'
    severity: PiErrorSeverity;              // indication of how serious the error is, default is 'To Do'
    constructor(message: string, element: PiElement | PiElement[], locationdescription: string, severity?: PiErrorSeverity) {
        this.message = message;
        this.reportedOn = element;
        this.locationdescription = locationdescription;
        if (!!severity) {
            this.severity = severity;
        } else {
            this.severity = PiErrorSeverity.ToDo;
        }
    }
}

export enum PiErrorSeverity {
    Error = "Error",
    Improvement = "Improvement",
    ToDo = "TODO",
    Info = "Info",
    NONE = "NONE"
}
