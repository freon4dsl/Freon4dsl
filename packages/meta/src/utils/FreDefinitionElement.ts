import { ParseLocation, FreParseLocation } from "./parsingAndChecking/FreGenericParser";

// root of the inheritance structure of all elements that can be present in any of the definitions
export class FreDefinitionElement {
    location: ParseLocation;        // used to store the location information from the PEGJS parser
    agl_location: FreParseLocation;  // used to store the location information from the AGL parser
}
