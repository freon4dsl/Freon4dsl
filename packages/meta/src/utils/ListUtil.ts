export class ListUtil {
    static addIfNotPresent<T>(list: T[], addition: T) {
        if (!!addition && !list.includes(addition)) {
            list.push(addition);
        }
    }

    static addListIfNotPresent<T>(list: T[], additionsList: T[]) {
        additionsList.forEach(extra => {
            ListUtil.addIfNotPresent<T>(list, extra);
        })
    }
}
