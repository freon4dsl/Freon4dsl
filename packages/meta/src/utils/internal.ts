export * from "./PiDefinitionElement";
export * from "./ListUtil";
export * from "./UserLogger";
export * from "./MetaLogger";
export * from "./parsingAndChecking";
export * from "./generation";
