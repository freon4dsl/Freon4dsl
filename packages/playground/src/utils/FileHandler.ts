import * as fs from "fs";

export class FileHandler {

    /**
     * Reads a string from the file located at 'filepath'. If the
     * file is not present an Error will be thrown.
     * @param filepath
     */
    public stringFromFile(filepath: string): string {
        // read language file
        if (!fs.existsSync(filepath)) {
            console.error(this, "File '" + filepath + "' does not exist, exiting.");
            throw new Error(`File '${filepath}' not found.`);
        }
        return fs.readFileSync(filepath, { encoding: "utf8" });
    }

    /**
     * Writes a string to the file located at 'filepath'. If the
     * file is not present it will be created.
     * May throw an Error if the file cannot be written or created.
     * @param filepath
     * @param output
     */
    public stringToFile(filepath: string, output: string) {
        if (fs.existsSync(filepath)) {
            console.log(this, "FileHandler: file " + filepath + " already exists, overwriting it.");
        }
        fs.writeFileSync(filepath, output);
    }

}
