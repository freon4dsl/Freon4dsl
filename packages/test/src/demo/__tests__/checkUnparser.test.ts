import { FreNodeReference } from "@freon4dsl/core";
import {
    DemoModel,
    DemoMultiplyExpression,
    DemoNumberLiteralExpression,
    DemoDivideExpression,
    DemoVariableRef,
    DemoEntity,
    DemoAttribute,
    DemoFunction,
    DemoVariable,
    DemoAttributeType
} from "../language/gen";
import { DemoModelCreator } from "./DemoModelCreator";
import { makeLiteralExp, MakeMultiplyExp, MakePlusExp } from "./HelperFunctions";
import { DemoValidator } from "../validator/gen";
import { DemoEnvironment } from "../config/gen/DemoEnvironment";
import { FileHandler } from "../../utils/FileHandler";


describe("Testing Unparser", () => {
    describe("Unparse DemoModel Instance", () => {
        const model: DemoModel = new DemoModelCreator().createIncorrectModel().models[0];
        const unparser = DemoEnvironment.getInstance().writer;

        beforeEach(done => {
            done();
        });

        test("3", () => {
            let result: string = "";
            const left = new DemoNumberLiteralExpression();
            left.value = 3;
            result = unparser.writeToString(left, 0);
            expect(result).toBe("3");
        });

        test("multiplication 3 * 10", () => {
            let result: string = "";
            const mult: DemoMultiplyExpression = new DemoMultiplyExpression();
            mult.left = makeLiteralExp("3");
            mult.right = makeLiteralExp("10");
            result = unparser.writeToString(mult, 0);
            expect(result).toBe("3 * 10");
        });

        test("multiplication 3 * 'temp'", () => {
            let result: string = "";
            const mult: DemoMultiplyExpression = new DemoMultiplyExpression();
            mult.left = makeLiteralExp("3");
            mult.right = makeLiteralExp("temp");
            result = unparser.writeToString(mult, 0);
            expect(result).toBe("3 * ' \"temp\" '");
        });

        test("multiplication 3 / 4 * 'temp'", () => {
            let result: string = "";
            const div: DemoDivideExpression = new DemoDivideExpression();
            div.left = makeLiteralExp("3");
            div.right = makeLiteralExp("4");
            const mult: DemoMultiplyExpression = new DemoMultiplyExpression();
            mult.left = div;
            mult.right = makeLiteralExp("temp");
            result = unparser.writeToString(mult, 0);
            expect(result).toBe("3 / 4 * ' \"temp\" '");
        });

        test("1 + 2 * 'Person'", () => {
            let result: string = "";
            const variableExpression = new DemoVariableRef();
            const variable = new DemoVariable();
            variable.name = "Person";
            // variable.declaredType = DemoAttributeType.String;
            variableExpression.variable = FreNodeReference.create<DemoVariable>(variable.name, "DemoVariable");

            const divideExpression = MakePlusExp("1", "2");
            const multiplyExpression = MakeMultiplyExp(divideExpression, variableExpression);
            result = unparser.writeToString(multiplyExpression, 0, false);
            result = result.replace(new RegExp("\\s+","gm"), " ");
            expect(result).toBe("1 + 2 * Person");
        });

        test('\'determine(AAP : TEST1) : TEST2 = "Hello Demo" + "Goodbye"\'', () => {
            let result: string = "";
            const determine = DemoFunction.create({ name: "determine" });
            const AAP = DemoVariable.create({ name: "AAP" });
            determine.parameters.push(AAP);
            AAP.declaredType = FreNodeReference.create<DemoEntity>(DemoEntity.create({name: "TEST1"}), "DemoEntity");
            determine.expression = MakePlusExp("Hello Demo", "Goodbye");
            determine.declaredType = FreNodeReference.create<DemoEntity>(DemoEntity.create({name: "TEST2"}), "DemoEntity");
            // determine(AAP: TEST1) : TEST2 = "Hello Demo" + "Goodbye" has been created
            // unparse using a short notation
            result = unparser.writeToString(determine, 0, true);
            expect(result).toBe("DemoFunction determine {");
            // unparse using a long notation
            result = unparser.writeToString(determine);
            expect(result).toMatchSnapshot();
        });

        test("Person { unitName, age, first(Resultvar): Boolean = 5 + 24 }", () => {
            let result: string = "";
            let myType = FreNodeReference.create<DemoAttributeType>(DemoAttributeType.Boolean, "DemoAttributeType");
            const personEnt = DemoEntity.create({ name: "Person" });
            const age = DemoAttribute.create({ name: "age", declaredType: myType });

            myType = FreNodeReference.create<DemoAttributeType>(DemoAttributeType.String, "DemoAttributeType");
            const personName = DemoAttribute.create({ name: "name", declaredType: myType });
            personEnt.attributes.push(age);
            personEnt.attributes.push(personName);
            myType = FreNodeReference.create<DemoAttributeType>(DemoAttributeType.Boolean, "DemoAttributeType");
            const first = DemoFunction.create({ name: "first", declaredType: myType });
            const myEntType = FreNodeReference.create<DemoEntity>("someOtherEntity", "DemoEntity");
            const resultvar = DemoVariable.create({ name: "Resultvar", declaredType: myEntType });
            first.parameters.push(resultvar);
            first.expression = MakePlusExp("5", "24");
            personEnt.functions.push(first);

            result = unparser.writeToString(personEnt, 0, false);
            // console.log(result)
            expect(result).toBe("DemoEntity Person {\n" +
                "    baseInterface_attr 0\n" +
                "    simpleprop \"\"\n" +
                "    x \"\"\n" +
                "    attributes\n" +
                "        age : Boolean\n" +
                "        name : String\n" +
                "    entAttributes\n" +
                "\n" +
                "    functions\n" +
                "        DemoFunction first {\n" +
                "            expression 5 + 24\n" +
                "            parameters\n" +
                "                Resultvar : someOtherEntity\n" +
                "            declaredType Boolean\n" +
                "        }\n" +
                "    int_attrs\n" +
                "\n" +
                "    int_functions\n" +
                "\n" +
                "}");
        });

        test("complete example model with simple attribute types", () => {
            let result: string = "";
            const testmodel = new DemoModelCreator().createModelWithMultipleUnits();
            const fileHandler = new FileHandler();

            const validator = new DemoValidator();
            const errors = validator.validate(testmodel, true);
            // errors.forEach(err =>{
            //    console.log((err.message + " in " + err.locationdescription))
            // });
            // the custom validation adds error message to an otherwise correct model
            expect(errors.length).toBe(9);

            for (const unit of testmodel.models) {
                result = unparser.writeToString(unit, 0, false);
                // fileHandler.stringToFile(`src/demo/__tests__/unparsed${unit.name}.txt`, result);

                // console.log(result);
                expect(result).toMatchSnapshot();
            }
        });
    });
});
