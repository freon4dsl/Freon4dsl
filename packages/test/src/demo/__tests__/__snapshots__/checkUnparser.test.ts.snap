// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Unparser Unparse DemoModel Instance 'determine(AAP : TEST1) : TEST2 = "Hello Demo" + "Goodbye"' 1`] = `
"DemoFunction determine {
    expression ' \\"Hello Demo\\" ' + ' \\"Goodbye\\" '
    parameters
        AAP : TEST1
    declaredType TEST2
}"
`;

exports[`Testing Unparser Unparse DemoModel Instance complete example model with simple attribute types 1`] = `
"model DemoModel_with_inheritance {
main:
types:

entities:
    DemoEntity Vehicle {
        baseInterface_attr 0
        simpleprop \\"simple\\"
        x \\"xxx\\"
        attributes
            brand : String
            type : String
        entAttributes

        functions
            DemoFunction SOME_VEHICLE {
                expression if ( 4 < 80 ) then
                                     87
                           else
                              1345
                           endif
                parameters

                declaredType Integer
            }
        int_attrs

        int_functions

    }
    DemoEntity Car {
        baseInterface_attr 0
        simpleprop \\"simple\\"
        x \\"xxx\\"
        attributes
            numberplate : Integer
            make : String
        entAttributes

        functions
            DemoFunction SOME_CAR {
                expression if ( 4 < 80 ) then
                                     87
                           else
                              1345
                           endif
                parameters

                declaredType Integer
            }
        int_attrs

        int_functions

        baseEntity Vehicle
    }
    DemoEntity Bike {
        baseInterface_attr 0
        simpleprop \\"simple\\"
        x \\"xxx\\"
        attributes
            backseat : Boolean
            gears : Integer
        entAttributes

        functions
            DemoFunction SOME_BIKE {
                expression if ( 4 < 80 ) then
                                     87
                           else
                              1345
                           endif
                parameters

                declaredType Integer
            }
        int_attrs

        int_functions

        baseEntity Vehicle
    }
    DemoEntity RaceBike {
        baseInterface_attr 0
        simpleprop \\"simple\\"
        x \\"xxx\\"
        attributes
            color : String
            wheelsize : Integer
        entAttributes

        functions
            DemoFunction SOME_RACEBIKE {
                expression if ( 4 < 80 ) then
                                     87
                           else
                              1345
                           endif
                parameters

                declaredType Integer
            }
        int_attrs

        int_functions

        baseEntity Bike
    }
model wide functions:
    DemoFunction SOME_MODEL {
        expression if ( 4 < 80 ) then
                             87
                   else
                      1345
                   endif
        parameters

        declaredType Integer
    }
}"
`;

exports[`Testing Unparser Unparse DemoModel Instance complete example model with simple attribute types 2`] = `
"model CorrectUnit {
main:
types:

entities:
    DemoEntity School {
        baseInterface_attr 0
        simpleprop \\"simple\\"
        x \\"xxx\\"
        attributes
            foundedIn : Integer
            name : String
        entAttributes

        functions
            DemoFunction requestClean {
                expression 5 + 24
                parameters
                    cleaningCompany : Company
                declaredType Integer
            }
        int_attrs

        int_functions

    }
    DemoEntity Company {
        baseInterface_attr 0
        simpleprop \\"simple\\"
        x \\"xxx\\"
        attributes
            name : String
            VAT_Number : Integer
        entAttributes

        functions
            DemoFunction doClean {
                expression 5 + 24
                parameters
                    at : Company
                declaredType Integer
            }
        int_attrs

        int_functions

    }
model wide functions:
    DemoFunction compare {
        expression if ( 2 < 5 ) then
                             1
                   else
                      5
                   endif
        parameters

        declaredType Integer
    } DemoFunction helloString {
          expression ' \\"Hello Demo\\" '
          parameters

          declaredType String
      }
}"
`;
